### PR TITLE
WIP: Feature/zfill

### PR DESCRIPTION
**Pull request by EelaNagarajQC**
**DRAFT** -- until this includes relevant updates to `StringBuilder` as contained in the PR for issue #160.

## Description
This PR addresses issue #134. Note: we follow pandas' convention for dealing with '+'/'-' (i.e. no special treatment) which is different from Python's `zfill` functionality.

### CODE REVIEW

File: fletcher/algorithms/string.py [hunk 1]
```diff
@@ -267,6 +268,79 @@ def _text_contains_case_sensitive(data: pa.Array, pat: str) -> pa.Array:
...
+        bit_offset = (row_idx + valid_offset) % 8
+        mask = np.uint8(1 << bit_offset)
+        valid = valid_bits[byte_offset] & mask
+
+        if not valid:
+            str_builder.append_null()
+            continue
+
+        val = data[row_idx]
+        value = np.concatenate(
```
Comment (EelaNagarajQC) [inline fletcher/algorithms/string.py:317]: @xhochy: the question in the draft refers to this part (and the identical part below): specifically the use of `as_py` and whether you have any general ideas/guidance on a better way of doing this. When we take the length of the raw data in the buffer, we run into issues (i.e. with some special characters whose raw representation is longer than 1 char but which should be counted as only a single character towards the string's length).

Comment (xhochy) [reply fletcher/algorithms/string.py:317]: Don't use `data` but rather extract the `value` from `data_buffer` using the offsets in `offsets`. To enable the usage of `numba`, you shouldn't need to pass in the `data: pa.Array,` argument here.

Comment (xhochy) [reply fletcher/algorithms/string.py:317]: A good trick to count the length of an UTF8-string is to count all bytes where the bit-pattern is not `11xxxxx`.

Comment (xhochy) [reply fletcher/algorithms/string.py:317]: A good trick to count the length of an UTF8-string is to count all bytes where the bit-pattern is not `10xxxxx`.

`11xxxxx` is the start of an UTF-8 character
`00xxxxx` is an ASCII character

Comment (xhochy) [reply fletcher/algorithms/string.py:317]: This should help you to get rid of any usage of `data` and instead use `data_buffer`.

Comment (EelaNagarajQC) [reply fletcher/algorithms/string.py:317]: thanks! will get started on that and will probably have some follow up questions later 😄

File: fletcher/algorithms/string.py [hunk 2]
```diff
@@ -267,6 +268,72 @@ def _text_contains_case_sensitive(data: pa.Array, pat: str) -> pa.Array:
...
+@njit
+def _get_zstring(width: int, buffer: np.ndarray) -> np.ndarray:
+    num_zeroes = width
+    for i in np.bitwise_xor(np.right_shift(buffer, 6), 2):
+        if i != 0:
+            num_zeroes -= 1
+    if num_zeroes <= 0:
+        return buffer
+    # Create array of byte-representation of 0s concat'd with buffer
+    return np.concatenate((np.full(num_zeroes, 48, dtype=np.uint8), buffer), axis=0)
```
Comment (xhochy) [inline fletcher/algorithms/string.py:280]: I would expect this line to be quite compute-heavy as you allocate space several times. A possibly more performant alternative would be to have a local buffer that is reused between each row. You only need to allocate that once as the upper bound of the size is `4 * width`. Any strings longer than that don't get zero-filled, smaller ones get zero-filled but won't be larger than `4 * width` after being zero-filled. 

With this trick, no more temporary allocations should be possible.

Comment (EelaNagarajQC) [reply fletcher/algorithms/string.py:280]: We were just about to ask you for some additional suggestions so will go ahead and work on this. Much appreciated! 😄 @ArseniiDunaevQC

File: fletcher/algorithms/string.py [hunk 3]
```diff
@@ -267,6 +268,72 @@ def _text_contains_case_sensitive(data: pa.Array, pat: str) -> pa.Array:
...
+
+
+@njit
+def _zfill_nonnull(
+    length: int, offsets: np.ndarray, data_buffer: np.ndarray, width: int
+):
+    str_builder = StringArrayBuilder(max(len(data_buffer), 2))
```
Comment (EelaNagarajQC) [inline fletcher/algorithms/string.py:287]: Also a style/intuition question @xhochy : is there a preference for passing in `str_builder` as a parameter of `_zfill_nonnull` and `_zfill_nulls` (i.e. initializing it within the calling function `_zfill`) vs. initializing it within each of these functions? Benchmarks don't appear to make a noticeable difference, and I do like eliminating the duplicated line by passing it in, but wasn't sure if there's a style preference for one over the other. Thanks!

Comment (xhochy) [reply fletcher/algorithms/string.py:287]: I would also prefer to pass it in from the outside.

File: fletcher/algorithms/string.py [hunk 4]
```diff
@@ -267,6 +268,118 @@ def _text_contains_case_sensitive(data: pa.Array, pat: str) -> pa.Array:
...
+        if i != 0:
+            num_zeros -= 1
+    if num_zeros <= 0:
+        return buffer
+    # Create array of byte-representation of 0s concat'd with buffer
+    for buff_idx in range(num_zeros):
+        temp_buffer[buff_idx] = 48
+    for buff_idx in range(num_zeros, num_zeros + len(buffer)):
+        temp_buffer[buff_idx] = buffer[buff_idx - num_zeros]
+    return temp_buffer[0 : buff_idx + 1]
```
Comment (EelaNagarajQC) [inline fletcher/algorithms/string.py:298]: Hmm..our several (very slightly varied, one below, another that directly the new values without looping) attempts at creating a local buffer that is reused doesn't seem to bring any performance speedups, which leads me to believe we are doing something dumb. Do any tips come to mind? Wondering if it's still a problem to return the slice as opposed to dealing with this differently. @xhochy

Comment (xhochy) [reply fletcher/algorithms/string.py:298]: I have checked out your code locally and for me `benchmarks.TimeSuite.time_zfill_ext` takes 16ms. Out of these 16ms, 12ms are spent on `str_builder.append_value` and `str_builder.append_null`. I sadly don't see any easy chance to speed this up. It seems like they occur the usual `jitclass` slowdown I have talked about in the beginning. Let's see whether #160 improves that a bit.

Comment (EelaNagarajQC) [reply fletcher/algorithms/string.py:298]: Bummer but sounds good, thanks. We will then focus more on cleanup + coverage for now then.

**Comment by xhochy** [comment]
This needs a rebase on top of #160 which hopefully adds a bit of performance but otherwise can be merged.

File: fletcher/algorithms/string.py (hunk 4 above)
Comment (ArseniiDunaevQC) [reply fletcher/algorithms/string.py]: Current (not final) benchmarks for `pandas` vs `fletcher`:

`[ 75.00%] ··· benchmarks.TimeSuite.time_zfill      10.1±0.1ms`
`[100.00%] ··· benchmarks.TimeSuite.time_zfill_ext  14.9±0.1ms`

**Comment by ArseniiDunaevQC** [comment]
Addresses #134.

**Comment by EelaNagarajQC** [comment]
Cool, will wait until #160 is merged to master, then rebase master, fix imports, etc. and see if it provides any improvements. (Unless you think #160 is done and it would be best if we go ahead and rebase now?)

**Comment by EelaNagarajQC** [comment]
@ArseniiDunaevQC @jonathanhaasqc @xhochy Very weird--rebased (at the time, up through PR 160, currently out of date). The drop-in replacement `StringArrayBuilder` has made the performance of `zfill` so slow that it now times out on benchmarks. (Can confirm that the logic still seems correct--passes all tests, and when going through the benchmark examples.) Any ideas? I can only assume it's something small/dumb, but with only a cursory look at the new StringBuilder implementation, I can't tell what's going on.

**Comment by ArseniiDunaevQC** [comment]
> @ArseniiDunaevQC @jonathanhaasqc @xhochy Very weird--rebased (master at the time, up through PR 160). The drop-in replacement `StringArrayBuilder` has made the performance of `zfill` so slow that it now times out on benchmarks. (Can confirm that the logic still seems correct--passes all tests, and when going through the benchmark examples.) Any ideas? I can only assume it's something small/dumb, but with only a cursory look at the new StringBuilder implementation, I can't tell what's going on.

Sorry for the slight delay @EelaNagarajQC, I just got back to work on it. Unfortunately, I can only agree as I experience the same behaviour: All tests pass, but the benchmarks (`time_zfill_ext` in particulary) fails because of timeout. Do you have any ideas/news about the issue?

**Pull request closed without merging**